DelineateIt: break ties in point snapping with flow accumulation

emlys opened the pull request:
# Description
Fixes #60 
Also made a couple of minor code style improvements.

# Checklist
- [x] Updated HISTORY.rst (if these changes are user-facing)
- [x] Updated the user's guide (if needed)

dcdenu4 commented on src/natcap/invest/delineateit/delineateit.py:
```diff
@@ -472,9 +477,11 @@ def snap_points_to_nearest_stream(points_vector_path, stream_raster_path_band,
             pixel values are ``1`` (indicating a stream pixel) or ``0``
             (indicating a non-stream pixel).
+        flow_accum_raster_path (string): A path to a flow accumulation raster
+            that is aligned with the stream raster.
```
Maybe a note that this is used to break ties?

dcdenu4 reviewed (commented):
Hey @emlys I just had a few questions. Thanks for looking them over!

emlys commented on requirements-dev.txt:
```diff
@@ -18,6 +18,7 @@ pypiwin32; sys_platform == 'win32'  # pip-only
 setuptools>=8.0
 Sphinx>=1.3.1,!=1.7.1
 sphinx-rtd-theme
+sphinx-reredirects
```
Yeah, just `release/3.10`.

emlys commented on src/natcap/invest/delineateit/delineateit.py:
```diff
@@ -571,22 +578,37 @@ def snap_points_to_nearest_stream(points_vector_path, stream_raster_path_band,
+            stream_flow_accums = flow_accum_array[row_indexes, col_indexes]
+
+            # 1d index of max value in flattened array
+            nearest_stream_index_1d = numpy.argmax(
+                (distance_array == distance_array.min()) *
+                stream_flow_accums  # weight by flow accum to break ties
+            )
```
It does return the first occurrence. So if there are two stream pixels that are equally near to the point, and also have the same flow accumulation, it would snap to the first one (first relative to that array window).
```
>>> a
array([[1, 2, 3],
       [6, 5, 6]])
>>> numpy.argmax(a)
3
```
I'm not sure if that answers your concern?

emlys commented:
@dcdenu4 I added to the docstring like you suggested. I also refactored a little to be more efficient so it only uses the flow accumulation if there's a tie to break.

dcdenu4 reviewed (approved):
Hey @emlys , this looks good to me. I can't quite remember my `argmax` confusion but when looking at the updates you made it makes sense to me now!
